pennylane/beta/tapes/rev.py

```diff
@@ -0,0 +1,204 @@
+    Some further post-processing of this gives the derivative
+    :math:`\frac{\partial}{\partial\theta} \langle \hat{O} \rangle` for any observable O.
+
+    The reversible approach is similar to backpropagation, but trades off extra computation for
+    enhanced memory efficiency. Where backpropagation caches the state tensors at each step during
+    a forward pass, the reversible method only caches the final pre-measurement state.
+
+    Compared to the parameter-shift rule, the reversible method can
+    be faster or slower, depending on the density and location of parametrized gates in a circuit
+    (circuits with higher density of parametrized gates near the end of the circuit will see a
+    benefit).
+    """
+
+    def _grad_method(self, idx, use_graph=True, default_method="A"):
+        return super()._grad_method(idx, use_graph=use_graph, default_method=default_method)
+
+    def _matrix_elem(self, vec1, obs, vec2, device):
+        r"""Computes the matrix element of an observable.
+
+        That is, given two basis states :math:`\mathbf{i}`, :math:`\mathbf{j}`,
+        this method returns :math:`\langle \mathbf{i} \vert \hat{O} \vert \mathbf{j} \rangle`.
+        Unmeasured wires are contracted, and a scalar is returned.
+
+        Args:
+            vec1 (array[complex]): a length :math:`2^N` statevector
+            obs (.Observable): a PennyLane observable
+            vec2 (array[complex]): a length :math:`2^N` statevector
+            device (.QubitDevice): the device used to compute the matrix elements
+        """
+        # pylint: disable=protected-access
+        mat = device._reshape(obs.matrix, [2] * len(obs.wires) * 2)
+        wires = obs.wires
+
+        vec1_indices = ABC[: device.num_wires]
+
+        obs_in_indices = "".join(ABC_ARRAY[wires.tolist()].tolist())
+        obs_out_indices = ABC[device.num_wires : device.num_wires + len(wires)]
+        obs_indices = "".join([obs_in_indices, obs_out_indices])
+
+        vec2_indices = reduce(
+            lambda old_string, idx_pair: old_string.replace(idx_pair[0], idx_pair[1]),
+            zip(obs_in_indices, obs_out_indices),
+            vec1_indices,
+        )
+
+        einsum_str = "{vec1_indices},{obs_indices},{vec2_indices}->".format(
+            vec1_indices=vec1_indices,
+            obs_indices=obs_indices,
+            vec2_indices=vec2_indices,
+        )
+
+        return device._einsum(einsum_str, device._conj(vec1), mat, vec2)
+
+    def jacobian(self, device, params=None, **options):
+        # The parameter_shift_var method needs to evaluate the circuit
+        # at the unshifted parameter values; the pre-rotated statevector is then stored
+        # self._state attribute. Here, we set the value of the attribute to None
+        # before each Jacobian call, so that the statevector is calculated only once.
+        self._state = None
+        return super().jacobian(device, params, **options)

```

Comment:
This is a pattern I've now had to repeat across the Qubit, CV, and Reversible tapes. All three tapes require the ability to determine when they are in the _first_ parameter gradient computation of a Jacobian, to avoid redundant device executions.

Ideally, we should just be able to check `if idx==0:`:

```python

        # loop through each parameter and compute the gradient
        for idx, (l, param_method) in enumerate(zip(p_ind, allowed_param_methods)):

            if param_method == "0":
                # Independent parameter. Skip, as this parameter has a gradient of 0.
                continue

            if (method == "best" and param_method[0] == "F") or (method == "numeric"):
                # finite difference method
                g = self.numeric_pd(l, device, params=params, **options)

            elif (method == "best" and param_method[0] == "A") or (method == "analytic"):
                # analytic method
                g = self.analytic_pd(l, device, params=params, **options)
```

_However_, if the first parameter has param_method `0`, then this check will fail; `idx=1` will instead be the first parameter gradient evaluation.

Can anyone think of a more elegant way of passing this knowledge to `analytic_pd`?

Reply:
Something to consider in the future, potentially before moving out of beta folder